localstack/services/sqs/provider.py

```diff
@@ -351,6 +351,9 @@ def update_visibility_timeout(self, receipt_handle: str, visibility_timeout: int
             if standard_message not in self.inflight:
                 raise MessageNotInflight()
 
+            # we also have to update last_received, since that is used to calculate the visibility deadline.
+            # last_received is not actually used to indicate "true" receipts via receive_message, so this is safe.
```

Comment:
As far as I can tell it is used for true receipts in queue.get() which is called in receive_message, at line 435. I think the main reason why this is still safe is because these actions seems to be mutually exclusive (after all, the visibility timeout acts as lock). Once received, the visibility timeout will prohibit all further receive_message attempts to receiving this message. And we only can change the visibility timeout after receiving a message (due to the required receipt handle). Finally, we do no care about the case where the visibility timeout expires, someone receives the message, and we call change_visibility afterwards, because we have "lost" in that case anyway.
tl;dr the comment might be misleading. IF I understood the involved mechanics correctly.

Reply:
that's a great explanation! maybe we should think about changing the code to make a clear separation. maybe add something like a "visibility deadline" that is calculated separately from the last_received timestamp